### ISSUE TITLE
Scale stroke width per zoom level

### ISSUE THREAD

**Issue by devemux86** [mapsforge/mapsforge#772]
Currently we scale the stroke width per zoom level in rendering (for zoom >12).

We can make that optional, preferably via render theme rules.

**Comment by Klaus-Tockloth** [comment]
(Negative) Example with dashed lines:

![bildschirmfoto 2016-03-28 um 12 25 07](https://cloud.githubusercontent.com/assets/5493128/14077219/c5d82af8-f4e9-11e5-9958-31df35813ccb.jpg)

![bildschirmfoto 2016-03-28 um 12 27 27](https://cloud.githubusercontent.com/assets/5493128/14077224/cfc83a58-f4e9-11e5-94f1-101de300e1b6.jpg)

Link: https://groups.google.com/forum/?fromgroups#!topic/mapsforge-dev/JShjzUWdW8I

Klaus

**Comment by devemux86** [comment]
I pushed a new `scale-stroke-width` option for area, circle, line, lineSymbol and pathText rules.

Essentially where we use now scaled stroke, keeping the default value as true for compatibility.

See example in render theme v5 [sample](https://github.com/mapsforge/mapsforge/blob/dev/mapsforge-samples-android/assets/renderthemes/rendertheme-v5.xml#L713) for highway=footway.

**Comment by Klaus-Tockloth** [comment]
scale-stroke-width=false: Does this mean that the width of the line isn't scaled?

Hmm, my expectation concerning this issue is, that beside the width, also the length of each line part (stroke or gap) is scaled.

Klaus

**Comment by devemux86** [comment]
We have 2 scale factors in render themes, one fixed and one variable:
- The fixed scale factor is based on device's DPI (on Android, on Java is 1) and is calculated on start. It affects **all** measured values in render theme.
- The variable scale factor (on top of the former) is proportional to zoom level (>12). It affects **only** the stroke widths (and their dependencies like `dy`) - nothing else. That means that we need to have rules per zoom level (groups) if we disable it, or else widths will be narrow.

The 2nd one is made optional via the `scale-stroke-width` option.

**Comment by devemux86** [comment]
>Hmm, my expectation concerning this issue is, that beside the width, also the length of each line part (stroke or gap) is scaled.

Are you sure? That's the stroke dash array affected only once by device's scale factor.
I understand (from code and my custom maps) that only the width is scaled additionally by zoom. That's why it does not appear nice on large zoom levels.
The best result is often via a rule per zoom, something difficult with large complex themes.

**Comment by Klaus-Tockloth** [comment]
What I want to achieve is something like this (example):
- A dashed line has a stroke length of 20 and a gap length of 10.
- In the next zoom level the width is increased by 10 percent.

Result: The stroke length has to be increased to 22 and the gap length to 11.

In other words: The scaling doesn't change the proportions of the line.

**Comment by devemux86** [comment]
Until now (and with default scale-stroke-width=true) in your example only the stroke width will change by zoom. The dashed line parameters remain the same by zoom.
So essentially you'd need a new rule in next zoom level (to modify all parameters).

**Comment by Klaus-Tockloth** [comment]
> So essentially you'd need a new rule in next zoom level (to modify all parameters).

You mean something like this:

```
        <rule e="way" k="*" v="feldweg023" zoom-min="14" zoom-max="14">
          <line stroke="#bd6500" stroke-width="0.6" stroke-dasharray="8,4" stroke-linecap="butt" />
        </rule>
        <rule e="way" k="*" v="feldweg023" zoom-min="15" zoom-max="15">
          <line stroke="#bd6500" stroke-width="0.6" stroke-dasharray="10,6" stroke-linecap="butt" />
        </rule>
        <rule e="way" k="*" v="feldweg023" zoom-min="16" zoom-max="16">
          <line stroke="#bd6500" stroke-width="0.6" stroke-dasharray="12,8" stroke-linecap="butt" />
        </rule>
        <rule e="way" k="*" v="feldweg023" zoom-min="17" zoom-max="17">
          <line stroke="#bd6500" stroke-width="0.6" stroke-dasharray="14,10" stroke-linecap="butt" />
        </rule>
        <rule e="way" k="*" v="feldweg023" zoom-min="18" zoom-max="18">
          <line stroke="#bd6500" stroke-width="0.6" stroke-dasharray="16,12" stroke-linecap="butt" />
        </rule>
        <rule e="way" k="*" v="feldweg023" zoom-min="19">
          <line stroke="#bd6500" stroke-width="0.6" stroke-dasharray="18,14" stroke-linecap="butt" />
        </rule>
```
... not very practical for a huge theme.

**Comment by devemux86** [comment]
Yes and with `scale-stroke-width=false` too. Indeed (having done that) not something trivial.

An alternative would be to have all rule numeric values scale with zoom (like stroke width) ?

**Comment by Klaus-Tockloth** [comment]
> Yes and with scale-stroke-width=false too.
Why is this useful if I change all parameter values in each zoom level?

> Indeed (having done that) not something trivial.
I agree.

> An alternative would be to have all rule numeric values scale with zoom (like stroke width) ?
Yes, that was exactly the intention of my feature request.

**Comment by severak** [comment]
Thanks for this change. When it propagates to renderers, it will make my rules a lot simpler.

BTW - today i am using lua [preprocessing script](https://github.com/severak/mapstyles/blob/master/metropolis/preprocess.lua) when developing themes.

**Comment by devemux86** [comment]
@Klaus-Tockloth zoom level scale influences always stroke width (if not disabled).

@severak thanks for the hint!

**Comment by devemux86** [comment]
For extending the scale to other directives too, I created #814.

**Comment by devemux86** [comment]
Closing this in favor of #814 and more generic implementation.
